src/bp/ui-admin/src/Pages/Server/Monitoring.jsx

```diff
@@ -0,0 +1,335 @@
+import React, { Component } from 'react'
+import { connect } from 'react-redux'
+import _ from 'lodash'
+import { IoIosBoxOutline } from 'react-icons/lib/io'
+import { Label, Row, Col, Container, Jumbotron } from 'reactstrap'
+import {
+  LineChart,
+  Line,
+  XAxis,
+  YAxis,
+  CartesianGrid,
+  Tooltip,
+  Legend,
+  Bar,
+  ComposedChart,
+  ResponsiveContainer
+} from 'recharts'
+
+import Select from 'react-select'
+import moment from 'moment'
+import ms from 'ms'
+import SummaryTable from '../Components/Monitoring/SummaryTable'
+import ChartTooltip from '../Components/Monitoring/ChartTooltip'
+import LoadingSection from '../Components/LoadingSection'
+
+import { groupEntriesByTime, mergeEntriesByTime, calculateOverviewForHost } from 'common/monitoring'
+import { fetchStats, refreshStats } from '../../reducers/monitoring'
+
+const timeFrameOptions = [
+  { value: '1m', label: '1 minute' },
+  { value: '5m', label: '5 minutes' },
+  { value: '10m', label: '10 minutes' },
+  { value: '15m', label: '15 minutes' },
+  { value: '30m', label: '30 minutes' },
+  { value: '1h', label: '1 hour' },
+  { value: '2h', label: '2 hours' },
+  { value: '5h', label: '5 hours' }
+]
+
+const resolutionOptions = [
+  { value: '10s', label: '10 seconds' },
+  { value: '30s', label: '30 seconds' },
+  { value: '1m', label: '1 minutes' },
+  { value: '5m', label: '5 minutes' },
+  { value: '15m', label: '15 minutes' },
+  { value: '30m', label: '30 minutes' },
+  { value: '1h', label: '1 hour' }
+]
+
+const tickSize = { fontSize: 11 }
+
+class Monitoring extends Component {
+  state = {
+    intervalId: null,
+    timeFrame: null,
+    resolution: null,
+    rawStats: null,
+    preparedStats: null,
+    timeFrameOptions
+  }
+
+  componentDidMount() {
+    this.setState({ timeFrame: timeFrameOptions[3], resolution: _.head(resolutionOptions) }, this.queryData)
+  }
+
+  componentDidUpdate(prevProps) {
+    if (prevProps.rawStats !== this.props.rawStats) {
+      this.prepareForDisplay()
+    }
+  }
+
+  async queryData() {
+    if (!this.state.timeFrame) {
+      return
+    }
+
+    try {
+      const fromTime = moment()
+        .subtract(ms(this.state.timeFrame.value))
+        .toDate()
+        .getTime()
+
+      const toTime = moment()
+        .toDate()
+        .getTime()
+
+      this.props.fetchStats(fromTime, toTime)
+    } catch (error) {
+      this.setState({ error: error.message })
+    }
+  }
+
+  prepareForDisplay() {
+    if (!this.props.rawStats || !this.props.rawStats.length) {
+      return
+    }
+
+    const lastUniqueEntries = _.sortBy(_.uniqBy(_.orderBy(this.props.rawStats, ['ts'], ['desc']), 'host'), 'host')
```

Comment:
I think I'd go with custom code here and _try_ to get your data on a single iteration. For each lodash function here it iterates over the array, so its pretty heavy.

Reply:
It would make it much more complicated. Basically I want to get the most recent entry for all unique hosts, and display them in alphabetical order. To get the list of unique hosts would still force a full iteration, then loop the array reversely until you get the most recent entry for each. 

I guess we could optimise when there is a big performance hit